src/dbnode/persist/fs/msgpack/decoder.go

```diff
@@ -711,3 +731,17 @@ func (dec *Decoder) decodeBytesLen() int {
 	dec.err = err
 	return value
 }
+
+func (dec *Decoder) wrapWithStreamWithDigest() {
+	dec.streamWithDigest.Reset(dec.reader)
+	dec.Reset(dec.streamWithDigest)
+}
+
+func (dec *Decoder) checksumAndUnwrapStreamWithDigest() uint32 {
```

Comment:
Is the intent of the wrapping and unwrapping logic at the decoder level to support decoding a series of non homogenous index entry versions?

Reply:
I think it's moreso to ensure we don't do any unnecessary work when decoding non-index entries (i.e. calculate useless checksums) and to also ensure that checksums are validated in a transparent fashion (i.e. without requiring decoder clients to know they need to do any validation). Also has the added benefit of being much easier to access the raw bytes for calculating the digest at this level.